### PR TITLE
GitLab CI: job generator checks if container images exists

### PR DESCRIPTION
**Pull request by SimeonEhrig**
- Fetch all images from the container registry when the jobs are generated.
- If an image does not exist use the best working alternative.
- The check can be disabled because it takes a lot of time and is annoying if you want to test a regex to filter jobs.

### CODE REVIEW

File: docs/source/dev/ci.rst [hunk 1]
```diff
@@ -132,6 +132,10 @@ One of them or none is also possible.
 
   You can test your regex offline before creating and pushing a commit. The ``job_generator.py`` provides the ``--filter`` and ``--reorder`` flags that do the same thing as the lines starting with ``CI_FILTER`` and ``CI_REORDER`` in the commit message.
 
+.. hint::
+
+  Each time the job generator runs, it checks whether the container images are exists. This is done by a request to the container registry. This takes a lot of time. Therefore, you can skip the check with the ``--no-image-check`` argument to speed up checking filter and reordering regex strings.
```
Comment (j-stephan) [inline docs/source/dev/ci.rst]:
```suggestion
  Each time the job generator runs it checks whether the container images exist. This is done by a request to the container registry which takes a lot of time. Therefore you can skip the check with the ``--no-image-check`` argument to speed up checking filters and reordering regex strings.
```

File: script/job_generator/generate_job_yaml.py [hunk 2]
```diff
@@ -70,44 +74,142 @@ def job_prefix_coding(job: Dict[str, Tuple[str, str]]) -> str:
 
 
 @typechecked
-def job_image(job: Dict[str, Tuple[str, str]], container_version: float) -> str:
-    """Generate the image url deppending on the host and device compiler and the selected backend
-    of the job.
+def get_images_from_registry(container_version: str) -> List[str]:
+    """Returns list of all container images for a given tag.
```
Comment (j-stephan) [inline script/job_generator/generate_job_yaml.py]:
```suggestion
    """Returns a list of all container images for a given tag.
```

File: script/job_generator/generate_job_yaml.py [hunk 3]
```diff
@@ -70,44 +74,142 @@ def job_prefix_coding(job: Dict[str, Tuple[str, str]]) -> str:
...
+
+
+@typechecked
+def job_image(
+    job: Dict[str, Tuple[str, str]],
+    container_version: float,
+    gitlab_images: List[str] = {},
+) -> str:
+    """Generate the image url depending on the host and device compiler and the
+    selected backend of the job.
```
Comment (j-stephan) [inline script/job_generator/generate_job_yaml.py]:
```suggestion
    """Generates the image URL depending on the host and device compiler and the
    job's selected back-end.
```

File: script/job_generator/generate_job_yaml.py [hunk 4]
```diff
@@ -70,44 +74,142 @@ def job_prefix_coding(job: Dict[str, Tuple[str, str]]) -> str:
...
+
+    If gitlab_images is not empty, only images that are exist are returned.
+    Otherwise, the name of the image is constructed by a set of rules. In this
+    case, the image may not exist. If it looks for an existing image and it does
+    not exist, it returns an image that best matches the software's requirements.
```
Comment (j-stephan) [inline script/job_generator/generate_job_yaml.py]:
```suggestion
    If gitlab_images is not empty only existing images are returned.
    Otherwise, the name of the image is constructed by a set of rules. However,
    the resulting image may not exist. In this case an image is returned that best
    matches the software's requirements.
```

File: script/job_generator/generate_job_yaml.py [hunk 5]
```diff
@@ -70,44 +74,142 @@ def job_prefix_coding(job: Dict[str, Tuple[str, str]]) -> str:
...
 
     Args:
         job (Dict[str, Tuple[str, str]]): Job dict.
         container_version (float): Container version tag.
+        gitlab_images (List[str]): List of existing container images. If it is
+            empty does not check if image exist.
```
Comment (j-stephan) [inline script/job_generator/generate_job_yaml.py]: This sentence seems incomplete.

File: script/job_generator/generate_job_yaml.py [hunk 6]
```diff
@@ -70,44 +74,142 @@ def job_prefix_coding(job: Dict[str, Tuple[str, str]]) -> str:
...
+    )
+
+    def verify_image(
+        test_url: List[str], verified_url: List[str], gitlab_images: List[str]
+    ) -> bool:
+        """Verify if the test_url is included in gitlab_images.
+
+        Args:
+            test_url (List[str]): Image URL to test. The list will concatenated
+                without whitespaces.
```
Comment (j-stephan) [inline script/job_generator/generate_job_yaml.py]: Is the list separated by whitespace?

File: script/job_generator/generate_job_yaml.py [hunk 7]
```diff
@@ -364,20 +466,31 @@ def create_job(
...
+    online_check: bool,
 ) -> List[Dict[str, Dict]]:
     """Generate the job yaml for each job in the job matrix.
 
     Args:
         job_matrix (List[List[Dict[str, Tuple[str, str]]]]): Job Matrix
         container_version (float): Container version tag.
+        online_check (bool): Check if image exist in the container registry.
+            Disabling the check can produce wrong job yaml's, but speed up the
+            job generation a lot of.
```
Comment (j-stephan) [inline script/job_generator/generate_job_yaml.py]:
```suggestion
        online_check (bool): Check if the image exists in the container registry.
            Disabling the check can produce wrong job YAMLs but speed up the
            job generation a lot.
```

File: script/job_generator/job_generator.py [hunk 8]
```diff
@@ -94,6 +95,12 @@ def get_args() -> argparse.Namespace:
         "--runtime-only", action="store_true", help="Generate only runtime jobs."
     )
 
+    parser.add_argument(
+        "--no-image-check",
+        action="store_false",
+        help="Disable check if Docker image exist in the registry.",
```
Comment (j-stephan) [inline script/job_generator/job_generator.py]:
```suggestion
        help="Disable registry check for existing Docker image.",
```

File: script/job_generator/job_generator.py [hunk 9]
```diff
@@ -183,7 +192,7 @@ def get_args() -> argparse.Namespace:
     wave_job_matrix = distribute_to_waves(job_matrix_yaml, {JOB_COMPILE_ONLY: 20})
 
     if wave_job_matrix[JOB_UNKNOWN]:
-        print('\033[33mWARNING: Generator distribute jobs of type "JOB_UNKNOWN"\033[m')
+        print_warn('Generator distribute jobs of type "JOB_UNKNOWN"')
```
Comment (j-stephan) [inline script/job_generator/job_generator.py]:
```suggestion
        print_warn('Generator distributed jobs of type "JOB_UNKNOWN"')
```

**Review by j-stephan — CHANGES_REQUESTED** [review]

File: docs/source/dev/ci.rst (hunk 1 above)
Comment (bernhardmgruber) [inline docs/source/dev/ci.rst]:
```suggestion
  Each time the job generator runs, it checks whether the container images exist. This is done by a request to the container registry. This takes a lot of time. Therefore, you can skip the check with the ``--no-image-check`` argument to speed up checking filter and reordering regex strings.
```

File: script/job_generator/generate_job_yaml.py [hunk 10]
```diff
@@ -70,44 +74,142 @@ def job_prefix_coding(job: Dict[str, Tuple[str, str]]) -> str:
...
+@typechecked
+def job_image(
+    job: Dict[str, Tuple[str, str]],
+    container_version: float,
+    gitlab_images: List[str] = {},
+) -> str:
+    """Generate the image url depending on the host and device compiler and the
+    selected backend of the job.
+
+    If gitlab_images is not empty, only images that are exist are returned.
```
Comment (bernhardmgruber) [inline script/job_generator/generate_job_yaml.py]:
```suggestion
    If gitlab_images is not empty, only images that exist are returned.
```

**Comment by SimeonEhrig** [comment]
[This PR](https://github.com/alpaka-group/alpaka/pull/2043) demonstrate how the PR works. The PR adds CUDA 12.2. It simply adds the version to the version.py and add a method to install the CUDA SDK during test time: https://github.com/alpaka-group/alpaka/pull/2043/commits/4785e3aa9705e5c561c72c8fd1d5214de39ef002 

The container registry does not provide an image for CUDA 12.2. Therefore the generator use a [base image](https://gitlab.com/hzdr/crp/alpaka/-/jobs/4781786197#L6) and install the CUDA SDK during test time: https://gitlab.com/hzdr/crp/alpaka/-/jobs/4781786197#L982

Compare to it, the CUDA 12.0 job takes the [CUDA 12.0 image](https://gitlab.com/hzdr/crp/alpaka/-/jobs/4781786223#L6) and [uses the preinstalled CUDA SDK](https://gitlab.com/hzdr/crp/alpaka/-/jobs/4781786223#L901).

File: script/job_generator/generate_job_yaml.py (hunk 6 above)
Comment (SimeonEhrig) [reply script/job_generator/generate_job_yaml.py]: The list is a python list. I use it, to add words to url at a specific position. For example:

```python
container_url=["registry.hzdr.de/crp/alpaka-group-container/alpaka-ci-ubuntu20.04", "-gcc"]

# concat without whitespace
print("".join(container_url))
# output: "registry.hzdr.de/crp/alpaka-group-container/alpaka-ci-ubuntu20.04-gcc"

container_url.insert(1, "-cuda")
# ["registry.hzdr.de/crp/alpaka-group-container/alpaka-ci-ubuntu20.04", "-cuda", "-gcc"]
print("".join(container_url))
# output: "registry.hzdr.de/crp/alpaka-group-container/alpaka-ci-ubuntu20.04-cuda-gcc"
```

Comment (j-stephan) [reply script/job_generator/generate_job_yaml.py]: Ah, I see. Then I'd rephrase it like this:

```suggestion
            test_url (List[str]): Image URL to test. The list will be concatenated
                with all whitespaces removed.
```

File: script/job_generator/generate_job_yaml.py [hunk 11]
```diff
@@ -70,44 +74,143 @@ def job_prefix_coding(job: Dict[str, Tuple[str, str]]) -> str:
...
+        test_url: List[str], verified_url: List[str], gitlab_images: List[str]
+    ) -> bool:
+        """Verify if the test_url is included in gitlab_images.
+
+        Args:
+            test_url (List[str]): Image URL to test. The list will concatenated
+                without whitespaces.
+            verified_url (List[str]): Working URL. Only for the warning message.
+                The Python list will concatenated to a string (URL) without
+                whitespaces or chars between the list elements.
```
Comment (j-stephan) [inline script/job_generator/generate_job_yaml.py]:
```suggestion
            verified_url (List[str]): Working URL. Only for the warning message.
                The Python list will be concatenated to a single string (URL) without
                whitespaces or chars between the list elements.
```

**Review by j-stephan — CHANGES_REQUESTED** [review]
Almost there :-)

**Review by j-stephan — APPROVED** [review]

**Comment by SimeonEhrig** [comment]
@j-stephan ready to merge
